File: Python/binary_search_02.py
# 22.01.18 떡볶이 떡 만들기
# 손님이 요청한 길이가 M 일 때, 최소 M 만큼(M 이상의) 떡을 손님이 받을 수 있는
# 절단기의 최댓값 구하기

# 절단기의 길이 : mid - mid 가 크면 총 길이가 짧아짐
# mid 가 작아지면 총 길이가 길어짐
# 총 길이가 요청 길이보다 짧으면 덜 잘라야함(mid의 크기가 줄어들어야함)
# 총 길이가 요청 길이보다 길면 더 잘라야함(mid 의 크기가 커져야함)

# n, m 입력
n, m = list(map(int, input().split(' ')))

# 떡의 개별 높이 입력
array = list(map(int, input().split()))
array.sort()

start = 0
end = max(array)
# 이진 탐색을 통해 최적의 절단기 높이 탐색
ans = None
while(start <= end):
    mid = (start + end) // 2
    total = 0
    # 잘랐을 때의 떡의 총 길이 계산
    for i in array:
        if i > mid:
            temp = i - mid
            total += temp
    # 총 길이가 부족한 경우 더 많이 절단(왼쪽 탐색)
    if total < m:
        end = mid - 1
    # 총 길이가 길 경우 덜 절단(오른쪽 탐색)
    else:
        ans = mid   # 최대한 덜 절단한 경우가 정답임
        start = mid + 1
        
print(ans)

File: Java/CodingTest/src/05_BinarySearch/BinarySearch_02.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

// 떡볶이 떡 만들기
public class BinarySearch_02 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        // N, M 입력
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        // 떡의 개별 높이 입력
        st = new StringTokenizer(br.readLine(), " ");
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        // 이진 탐색을 위한 정렬
        Arrays.sort(arr);

        // 이진 탐색 수행 후 결과 출력
        System.out.println(binarySearch(arr, m, 0, arr[n - 1]));
    }

    public static int binarySearch(int[] arr, int target, int start, int end) {
        int max = 0;
        while (start <= end) {
            int mid = (int) ((start + end) / 2);
            int sum = 0;
            for (int i = 0; i < arr.length; i++) {
                // 떡의 길이가 현재 절단기의 길이보다 길면 남고, 짧으면 남지 않음
                if (arr[i] > mid) {
                    sum += arr[i] - mid;
                }
            }
            // 요청한 떡의 길이보다 적을 경우 왼쪽 탐색(절단기 길이가 너무 긴 경우)
            if (sum < target) {
                end = mid - 1;
            }
            // 요청한 떡의 길이보다 크거나 같을 경우(절단기 길이가 적당한 경우)
            // 최댓값 갱신 후 오른쪽 탐색
            else {
                max = mid;
                start = mid + 1;
            }
        }
        return max;
    }
}
